## Тема: Описательная статистика. Качественные и количественные характеристики популяции. Графическое представление данных

In [45]:
import pandas as pd
import numpy as np
from math import factorial

# функция расчета количества сочетаний
def com(k, n):
    return int(factorial(n) / (factorial(k) * factorial(n - k)))

### 1. Даны значения зарплат из выборки выпускников: 100, 80, 75, 77, 89, 33, 45, 25, 65, 17, 30, 24, 57, 55, 70, 75, 65, 84, 90, 150. Посчитать (желательно без использования статистических методов наподобие std, var, mean) среднее арифметическое, среднее квадратичное отклонение, смещенную и несмещенную оценки дисперсий для данной выборки.

In [15]:
salary_lst = [100, 80, 75, 77, 89, 33, 45, 25, 65, 17, 30, 24, 57, 55, 70, 75, 65, 84, 90, 150]

In [16]:
df = pd.DataFrame({'salary':salary_lst})

**среднее арифметическое**

In [17]:
mean_salary = df['salary'].sum() / df['salary'].count()
mean_salary

65.3

**среднее квадратичное отклонение**

In [60]:
std_salary = np.round(np.sqrt(((df['salary'] - df['salary'].mean())**2).sum() / df['salary'].count()), 4)
std_salary

30.8239

**смещенная оценка дисперсии**

In [40]:
var0_salary = ((df['salary'] - df['salary'].mean())**2).sum() / df['salary'].count()
var0_salary

950.11

**несмещенная оценка дисперсии**

In [61]:
var1_salary = np.round(((df['salary'] - df['salary'].mean())**2).sum() / (df['salary'].count() - 1), 4)
var1_salary

1000.1158

### 2. В первом ящике находится 8 мячей, из которых 5 - белые. Во втором ящике - 12 мячей, из которых 5 белых. Из первого ящика вытаскивают случайным образом два мяча, из второго - 4. Какова вероятность того, что 3 мяча белые?

возможны 3 комбинации несовместных событий

1. A1 = {из 1-го ящика 0 белых, 2 черных} И B1 = {из 2-го ящика 3 белых, 1 черный} - совсестное событие 
2. A2 = {из 1-го ящика 1 белый, 1 черный} И B2 = {из 2-го ящика 2 белых, 2 черных} - совсестное событие  
3. A3 = {из 1-го ящика 2 белых, 0 черных} И B3 = {из 2-го ящика 1 белый, 3 черных} - совсестное событие 

Произведение первых совместных событий $P(AB) = P(A) * P(B) = \frac{C_5^0 * C_3^2}{C_8^2} * \frac{C_5^3 * C_7^1}{C_{12}^4}$

In [62]:
A1B1 = np.round((com(0,5) * com(2,3) / com(2,8)) * (com(3,5) * com(1,7) / com(4,12)), 4)
A1B1

0.0152

Произведение вторых совместных событий $P(AB) = P(A) * P(B) = \frac{C_5^1 * C_3^1}{C_8^2} * \frac{C_5^2 * C_7^2}{C_{12}^4}$

In [63]:
A2B2 = np.round((com(1,5) * com(1,3) / com(2,8)) * (com(2,5) * com(2,7) / com(4,12)), 4)
A2B2

0.2273

Произведение третьих совместных событий $P(AB) = P(A) * P(B) = \frac{C_5^2 * C_3^0}{C_8^2} * \frac{C_5^1 * C_7^3}{C_{12}^4}$

In [64]:
A3B3 = np.round((com(2,5) * com(0,3) / com(2,8)) * (com(1,5) * com(3,7) / com(4,12)), 4)
A3B3

0.1263

Сумма совсестных событий $P(A + B) = P(A) + P(B)$

In [65]:
P = A1B1 + A2B2 + A3B3
P

0.3688

### 3. На соревновании по биатлону один из трех спортсменов стреляет и попадает в мишень. Вероятность попадания для первого спортсмена равна 0.9, для второго — 0.8, для третьего — 0.6. Найти вероятность того, что выстрел произведен:

### a) первым спортсменом

Событие **А** = {попадание в мишень}  
События **В1, В2, В3** = {выстрел совершил первый, второй, третий спортсмен}  
**P(В1) = P(В2) = P(В3)** = 1/3  
**P(A|B1)** = 0.9  
**P(A|B2)** = 0.8  
**P(A|B3)** = 0.6  

In [66]:
B1 = B2 = B3 = 1/3
A_B1 = 0.9
A_B2 = 0.8
A_B3 = 0.6

Найдем вероятнось события **P(А)** по формуле **полной вероятности**: 

$$P(A) = P(B1) * P(A|B1) + P(B2) * P(A|B2) + P(B3) * P(A|B3) = 1/3 * 0.9 + 1/3 * 0.8 + 1/3 * 0.6$$

In [67]:
A = np.round((B1 * A_B1) + (B2 * A_B2) + (B3 * A_B3), 4)
A

0.7667

Найдем вероятность попадания первым спортсменом при условии, что выстрел был удачным по формуле **Байеса**: 

$$P(B|A) = \frac{P(B)*P(A|B)}{P(A} \Rightarrow P(B1|A) = \frac{P(B1) * P(A|B1)}{P(A)} = \frac{\frac{1}{3} * 0.9}{0.7666}$$

In [68]:
B1_A = np.round(B1 * A_B1 / A, 4)
B1_A

0.3913

### б) вторым спортсменом

$$P(B2|A) = \frac{P(B2) * P(A|B2)}{P(A)} = \frac{\frac{1}{3} * 0.8}{0.7666}$$

In [69]:
B2_A = np.round(B2 * A_B2 / A, 4)
B2_A

0.3478

### в) третьим спортсменом

$$P(B3|A) = \frac{P(B3) * P(A|B3)}{P(A)} = \frac{\frac{1}{3} * 0.6}{0.7666}$$

In [70]:
B3_A = np.round(B3 * A_B3 / A, 4)
B3_A

0.2609

### 4. В университет на факультеты A и B поступило равное количество студентов, а на факультет C студентов поступило столько же, сколько на A и B вместе. Вероятность того, что студент факультета A сдаст первую сессию, равна 0.8. Для студента факультета B эта вероятность равна 0.7, а для студента факультета C - 0.9. Студент сдал первую сессию. Какова вероятность, что он учится: 

### a) на факультете A?

Событие **А** = {сдача сессии}  
События **В1, В2, В3** = {сдал сессию студент факультета А, В, С}  
**P(В1) = P(В2)** = 1/4  
**P(В3)** = 2/4  
**P(A|B1)** = 0.8  
**P(A|B2)** = 0.7  
**P(A|B3)** = 0.9

In [71]:
B1 = B2 = 1/4
B3 = 2/4
A_B1 = 0.8
A_B2 = 0.7
A_B3 = 0.9

Найдем вероятнось события **P(А)** по формуле **полной вероятности**: 

$$P(A) = P(B1) * P(A|B1) + P(B2) * P(A|B2) + P(B3) * P(A|B3) = 1/4 * 0.8 + 1/4 * 0.7 + 2/4 * 0.9$$

In [72]:
A = (B1 * A_B1) + (B2 * A_B2) + (B3 * A_B3)
A

0.825

Найдем вероятность сдачи сессии студентом факультета А при условии, что сессия сдана по формуле **Байеса**: 

$$P(B|A) = \frac{P(B)*P(A|B)}{P(A} \Rightarrow P(B1|A) = \frac{P(B1) * P(A|B1)}{P(A)} = \frac{\frac{1}{4} * 0.8}{0.825}$$

In [73]:
B1_A = np.round(B1 * A_B1 / A, 4)
B1_A

0.2424

### б) на факультете B?

$$P(B1|A) = \frac{P(B1) * P(A|B1)}{P(A)} = \frac{\frac{1}{4} * 0.7}{0.825}$$

In [74]:
B2_A = np.round(B2 * A_B2 / A, 4)
B2_A

0.2121

 ### в) на факультете C?
 
 $$P(B1|A) = \frac{P(B1) * P(A|B1)}{P(A)} = \frac{\frac{2}{4} * 0.9}{0.825}$$

In [75]:
B3_A = np.round(B3 * A_B3 / A, 4)
B3_A

0.5455

### 5. Устройство состоит из трех деталей. Для первой детали вероятность выйти из строя в первый месяц равна 0.1, для второй - 0.2, для третьей - 0.25. Какова вероятность того, что в первый месяц выйдут из строя: 

### а) все детали 

**P(B1)** = 0.1  
**P(B2)** = 0.2  
**P(B3)** = 0.25

Т.к. события должны произойти одновременно используем формулу произведения: 

$$P(A) = P(B1) * P(B2) * P(B3) = 0.1 * 0.2 * 0.25$$

In [87]:
B1 = 0.1
B2 = 0.2
B3 = 0.25

A1 = np.round(B1 * B2 * B3, 4)
A1

0.005

### б) только две детали

возможны 3 комбинации несовместных событий  

1. B1 * B2  
2. B1 * B3  
3. B2 * B3  

Используем одновременно формулу сложения совместный событий и формулу произведения несовместных событий.

In [88]:
A2 = B1 * B2 + B1 * B3 + B2 * B3
A2

0.095

### в) хотя бы одна деталь 

Для упрощения расчетов можно использовать противоположные события

In [89]:
A3 = (1 - B1) * (1 - B2) * (1 - B3)
A3

0.54

### г) от одной до двух деталей

Для выполнения условия достаточно сложить уже расчитанные "только две детали" и вероятность одной детали.

In [90]:
A4 = A2 + B1 + B2 + B3
A4

0.645